plugins/application/prometheus/expiry.go

```diff
@@ -0,0 +1,60 @@
+package main
+
+import (
+	"container/list"
+	"context"
+	"time"
+)
+
+//expire metrics after time of staleness
+
+type expiry interface {
+	Expired() bool
+	Delete()
+}
+
+type expiryProc struct {
+	entries  *list.List
+	interval time.Duration
+}
+
+func newExpiryProc() *expiryProc {
+	return &expiryProc{
+		entries:  list.New(),
+		interval: 10000,
```

Comment:
Expire time might need to be based on the individual metric's sample interval.  I believe if the expire interval is set to 10 seconds and a variable's collectd interval is 60 seconds, there code be issues with variables expiring before they are sampled by prometheus.  The prometheus sample interval should be set to 60 seconds.  It is possible that a variable will get set and expire before it is sampled and then the timeseries will go invalid inside Prometheus.  

This brings up the issue of putting all metrics of the same interval into their own scrape target so that Prometheus can poll at different rates.  For example, disk-related metrics only need to be sample every 10 minutes (as an example), but packet counts every 30 seconds.  With a single scrape target, Prometheus would
need to scrape every 30 seconds thereby oversampling the disk metrics.  The solution is to have bins of scrape targets grouped by collectd interval.  The collectd interval is contained in each metric that is sent.

:)

Reply:
This configuration refers to how often the expiry process checks registered metrics to see if they are expired, not the actual interval for metrics: https://github.com/pleimer/sg-core/blob/ec42d56751897d12c991af34a400f43abefaf624/plugins/application/prometheus/expiry.go#L55

Currently a metric is marked as stale if one does not arrive in the timeout interval specified in the Prometheus Plugin config: https://github.com/pleimer/sg-core/blob/ec42d56751897d12c991af34a400f43abefaf624/plugins/application/prometheus/main.go#L239. By default this is 20s: https://github.com/pleimer/sg-core/blob/ec42d56751897d12c991af34a400f43abefaf624/plugins/application/prometheus/main.go#L168. Currently that configured timeout applies to all incoming metrics, but it would be simple to use an interval specified by the metric itself. I definitely see the utility in batching metrics according to timeout and seems like functionality we should have, I will add that to my backlog. That is not what happens in the current sg-core, however. Thanks for the feedback and ideas!

Reply:
Correct.  Adding time-based bins would be an enhancement.   However, the stale interval should really be based on individual metrics.  This was a glaring problem with the original implementation.